### ISSUE TITLE
Heliocentric JD converter shows wrong number of converted observations

### ISSUE THREAD

**Issue by mpyat2** [AAVSO/VStar#185]
Heliocentric JD converter always shows TOTAL number of observations as 'converted' (including observations that were HJD before conversion). Probably it should display the number of transformed observations only (in the datasets that were JD before conversion)?

**Comment by dbenn** [comment]
I'm trying to reproduce this one @mpyat2. If I load ASAS data for R Car then add to this AID vis and V data (375) for R Car then do HJD conversion, it correctly reports 375.

Can you give me another specific example to try?

I can see some code that if it was invoked under some circumstances could lead to what you observe, although I don't think it can be. I have a fix for this if it turns out to be the problem.

Let's start with your example scenario and go from there though.

**Comment by mpyat2** [comment]
Hi @dbenn ,
I was wrong: there is another bug: JDE converted converts BJD observations.
I've corrected the title and the description,

**Comment by mpyat2** [comment]
in HJDConverter.java:
line 69: should test isBarycentric() too
line 154: should test isBarycentric() too 
(like in lines 108-109)

**Comment by dbenn** [comment]
Yes, that makes sense and I completely @mpyat2.

Thanks for looking at this. I've been a bit busy lately.

Do you want to make these changes and I will review or would you like me to make the changes?

**Comment by mpyat2** [comment]
Hi David,
I'll try to fix it yet now I see an additional strange behavior. I'll try to sort it out and write more.

**Comment by mpyat2** [comment]
Hi David @dbenn 

The problem is somewhat deeper and is connected with isHeliocentric usage in the VStar code.
If I load JD dataset (from AAVSO DB, for example) after Barycentric (Kepler/TESS source) and try to convert to HJD, the BJD dataset also gets converted even with the code fixes.

I suspect this is because ValidObservation object is not aware of Barycentric mode.

**Comment by dbenn** [comment]
Indeed. The retriever class knows about HJD and BJD but ValidObservation knows only about HJD. I don't know why I allowed this asymmetry to exist. Adding isBarycentric() would make sense with similar usage patterns to isHeliocentric().

Let me know if you want help with this or just review.

**Comment by dbenn** [comment]
Ultimately, abstracting the notion of time with a canonical internal representation and allowing arbitrary transformations would be best, but I think that's a larger goal.

**Comment by mpyat2** [comment]
In general, the plugin logic is not worked as expected. The observations are not divided among retrievers, in fact, they are allin the last ValidObservations array. So I'll will do several fixes in different places to make it work.

**Comment by mpyat2** [comment]
Other related bugs mentioned in my previous comment I intend to fix:
1) If HeliocentricJD dataset loaded AFTER non-heliocentric (JD) dataset, Heliocentric JD converted tells: "No datasets with Julian Date observations". To reproduce, load any star from AAVSO database (JD) then add ASAS-SN dataset (HJD).
2) In "Observation details" dialog: try to load HJD dataset (ASAS-SN, for example), Click on an observation and invoke "Observation details". You will see "HJD" as a unit of measure. Then add JD dataset (from AAVSO database, for example), All ASAS-SN observation displayed as "JD" now in "Observation details". Then add BJD dataset (TESS, HIPPARCOS...). Now all observations are shown as BJD (in "Observation details")

**Comment by dbenn** [comment]
Yes, given that a list of retrievers is stored, I should have taken the extra step of keeping obs isolated within a retriever object and provided an aggregator observation view for plot and table models. Apologies that this is more complex than it should be @mpyat2. It's one aspect of the design I am least happy with and is least habitable.

**Comment by mpyat2** [comment]
No problem @dbenn . I made some steps, it seems the converter will work eventually. I hope I'll upload fixes soon.

**Comment by dbenn** [comment]
> @dbenn, probably yes (concerning abstracting the notion of time).
> 
> Yet now I think it would be worth getting rid of isBarycentric/isHeliocentric private fields, they could be replaced by an ENUM, something like timeFlavour, or something else. Getters/Setters can be remained as they are (isHeliocentric()/setHeliocentric(), probably isBarycentric()/setBarycentric() where appropriate).
> 
> In ValidObservation class, after introducing ENUM that defines JD/HJD/BJD, hashCode() should be fixed. Unfortunately, I've never made this before (creating proper hashCode() method), could you help with it? Equals() should be changed too, it is trivial.
> 
> I could play with these changes. I'm a little busy now yet probably will find a time.

Hi @mpyat2. No worries. If you are using eclipse, the simplest way to generate hashCode() and equals() is by ctrl-click anywhere in a class source file, selecting Source and then Generate hashCode() and equals().

**Comment by mpyat2** [comment]
Hi David @dbenn ,
Could you please review a new branch issue-185-HJDconverted-converts-BJD ?
I have not created a pull request because I'm not sure I finalized the work however I think the converted works better now.

A brief description of changes:
* HJDConverter plug-in modified to allow the correct transformation of JD observations (and somewhat simplified). This required changes in VStar core so the plugin is not compatible with the current VStar 2.21.3.
* VStar version changed to 2.21.4dev.
* JulianDateObservationsFilter plugin added to highlight existing JD (not converted to HJD) observation.

**Comment by dbenn** [comment]
This looks awesome @mpyat2! BTW, given the core change perhaps we can move to 2.22.0dev?

**Comment by mpyat2** [comment]
Good suggestion, @dbenn . I've updated the version number to 2.22.0dev
